src/lib/input/input-wrapper.ts

```diff
@@ -0,0 +1,289 @@
+let nextUniqueId = 0;
+
+
+export class MdInputWrapperInputElementCountError extends MdError {
+  constructor(count: number) {
+    super(`md-input-wrapper must contain exactly 1 input or textarea element. Found ${count}.`);
+  }
+}
+
+export class MdInputWrapperPlaceholderConflictError extends MdError {
+  constructor() {
+    super('Placeholder attribute and child element were both specified.');
+  }
+}
+
+export class MdInputWrapperUnsupportedTypeError extends MdError {
+  constructor(type: string) {
+    super(`Input type "${type}" isn't supported by md-input-wrapper.`);
+  }
+}
+
+export class MdInputWrapperDuplicatedHintError extends MdError {
+  constructor(align: string) {
+    super(`A hint was already declared for 'align="${align}"'.`);
+  }
+}
+
+/**
+ * The placeholder directive. The content can declare this to implement more
+ * complex placeholders.
+ */
+@Directive({
+  selector: 'md-placeholder'
+})
+export class MdPlaceholder {}
+
+
+/** The hint directive, used to tag content as hint labels (going under the input). */
+@Directive({
+  selector: 'md-hint',
+  host: {
+    '[class.md-right]': 'align == "end"',
+    '[class.md-hint]': 'true'
+  }
+})
+export class MdHint {
+  // Whether to align the hint label at the start or end of the line.
+  @Input() align: 'start' | 'end' = 'start';
+}
+
+
+/**
+ * Component that represents a text input. It encapsulates the <input> HTMLElement and
+ * improve on its behaviour, along with styling it according to the Material Design.
+ */
+@Component({
+  moduleId: module.id,
+  selector: 'md-input-wrapper',
+  templateUrl: 'input-wrapper.html',
+  styleUrls: ['input.css', 'input-wrapper.css'],
+  host: {
+    '(click)' : '_focusInput()',
+    // remove align attribute to prevent it from interfering with layout.
+    '[attr.align]': 'null',
+  },
+  encapsulation: ViewEncapsulation.None,
+})
+export class MdInputWrapper implements AfterContentInit, OnChanges {
+  @Input() align: 'start' | 'end' = 'start';
+
+  @Input() dividerColor: 'primary' | 'accent' | 'warn' = 'primary';
+
+  @Input() hintLabel: string = '';
+
+  @Input()
+  get floatingPlaceholder(): boolean { return this._floatingPlaceholder; }
+  set floatingPlaceholder(value) { this._floatingPlaceholder = coerceBooleanProperty(value); }
+  private _floatingPlaceholder: boolean = true;
+
+  @ContentChild(MdPlaceholder) _placeholderChild: MdPlaceholder;
+
+  @ContentChildren(MdHint) _hintChildren: QueryList<MdHint>;
+
+  @ContentChild(NgModel) _ngModelChild: NgModel;
+
+  /** Whether the `input` or `textarea` is focused. */
+  _focused = false;
+
+  /** The disabled attribute of the `input` or `textarea`. */
+  _inputDisabled = false;
+
+  /** The id attribute of the `input` or `textarea`. */
+  _inputId = '';
+
+  /** The required attribute of the `input` or `textarea`. */
+  _inputRequired = false;
+
+  /** Whether the `input` or `textarea` is empty. */
+  get _empty(): boolean { return !this._inputValue && this._inputType !== 'date'; }
```

Comment:
This won't necessarily be true on browsers that don't support those types, though. E.g. only Chrome supports `type="date"`. That's why I went with the feature test on #1860.